### PR TITLE
Allow re-using name-input layouts for different languages

### LINKED ISSUE #1272: Ib fan translation uses the Japanese text entry field instead of the English one
#### Name of the game:
Ib (Most recent English translation by vgperson)

#### Player platform:
Windows

#### Attach files (as a .zip archive or link them)
[easyrpg_log.txt](https://github.com/EasyRPG/Player/files/1361426/easyrpg_log.txt)
[Save01.zip](https://github.com/EasyRPG/Player/files/1361432/Save01.zip)

#### Describe the issue in detail and how to reproduce it:
In the "Secret Room" of The Dungeon in Ib, there's a room where you have to enter a name in order to proceed. With the standard RTP, the name entry uses English characters, allowing you to enter it, but 
with EasyRPG, it uses the Japanese name entry instead.

To reproduce:
- Load the attached save
- Enter the northwestern room
- Answer "Sure" at the dialogue prompt, then speak with Garry (the man) three times
- Wait for the objects in the room to start shaking, examine the newly uncovered canvas in the northeast part of the room, then examine the northwesternmost bookshelf
- Choose "Read how to enter"
- Examine the back of the leftmost bookcase in the southeast part of the room
- Stand between the bookcase and Garry, face west, then examine the gap
- Examine the canvas, and choose "Close eyes, count to three"
- Pull the cord that appears
- Examine the rightmost orb and choose "Enter a word"

### PR DESCRIPTION
**Pull request by rohkea**
Seems to work for me, please test.

Feedback is welcome. I’m particularly not happy with the forward declaration of ``struct Keyboard_Layout;``.

---
EDIT by @carstene1ns: Fixes #1272.

### CODE REVIEW

**Comment by Ghabry** [comment]
Test case:

https://easyrpg.org/play/pr1284/?start-map-id=36&start-position=6%2011&new-game

To test: via &encoding= : 932, 949, 936, 1251, 1252

**Comment by Ghabry** [comment]
When you use encoding 1252 and talk to the right-most event (which launches name scene on katakana page/ page 2) the continue button is not rendered.

**Comment by rohkea** [comment]
Thanks for noticing! It should be OK now.

**Comment by rohkea** [comment]
Thanks! I’ve changed int’s to size_t’s as suggested in the IRC, and reordered the initialisations.

**Comment by carstene1ns** [comment]
btw. one thing that has annoyed me since ever (mostly noticeable in the symbol page):
When you press down and there is no symbol under it, it will jump one to the left, but only when there is one, otherwise it jumps to the top or skips the empty row.

Demonstration: Press <kbd>Down</kbd> here.

screenshot

**Comment by Ghabry** [comment]
Which behaviour would you prefer? That it always jumps at ⇖ in the next row or that it goes downwards and jumps to the top?

**Comment by carstene1ns** [comment]
IMO it should never jump left anyway.

**Comment by carstene1ns** [comment]
Added a commit that fixes this behaviour.
It was actually needed to support moving down on the wide fields.
Achieving this differently now (*hack alert*).

File: src/window_keyboard.cpp [hunk 1]
```diff
@@ -16,20 +16,21 @@
...
-const char* const Window_Keyboard::DONE_RU = "<OK>";
-const char* const Window_Keyboard::DONE_KO = "<결정>";
-const char* const Window_Keyboard::DONE_ZH_CN = "<确定>";
+const char* Window_Keyboard::DONE = "<Done>";
+const char* Window_Keyboard::SPACE = "SPACE";
+const char* Window_Keyboard::NEXT_PAGE = "<Page>";
+const char* Window_Keyboard::DONE_JP = "<決定>";
+const char* Window_Keyboard::DONE_RU = "<OK>";
+const char* Window_Keyboard::DONE_KO = "<결정>";
+const char* Window_Keyboard::DONE_ZH_CN = "<确定>";
```
Comment (Ghabry) [inline src/window_keyboard.cpp]: How about "constexpr char Window_Keyboard::DONE[] =" ;)?

File: src/window_keyboard.cpp [hunk 2]
```diff
@@ -237,42 +238,37 @@ void Window_Keyboard::Refresh() {
...
-			}
+			col = (col + col_max - 1) % col_max;
 		}
+	}
 
+	// Special handling for wide fields
+	if (col > 0 && col < col_max) {
+		std::string k = GetKey(row, col - 1);
+		if (k[0] == '<' || k == SPACE)
+			col = std::min(col + skip_dir, col_max - 1);
```
Comment (Ghabry) [inline src/window_keyboard.cpp]: Is it possible check for SYMBOL and DONE based on the row-col position instead?
Imho checking for "<" is a bit ugly.

Comment (carstene1ns) [reply src/window_keyboard.cpp]: Yeah, my intention was here to not have to compare with all the different DONEs. But you are right, as long as they stay on their position your idea is good.

File: src/window_keyboard.cpp (hunk 1 above)
Comment (carstene1ns) [reply src/window_keyboard.cpp]: This does not work. Either I get `-Wwriteable-strings` warnings or the build errors out with
`error: non-const static data member must be initialized out of line` or
`error: declaration of constexpr static data member 'SPACE' requires an initializer` or
`error: redefinition of 'SPACE' with a different type: 'char *const' vs 'const char *'`, depending on how i modif/move them. 🤕

Comment (Ghabry) [reply src/window_keyboard.cpp]: o.O okay leave it as is.

**Pull request merged**